[ActionBar] Support mixed items (#317)
primer/design: content/components/action-bar.mdx

---
title: Action bar
status: Experimental
description: An action bar contains a collection of horizontally aligned actions.
---

import {Box, Text} from '@primer/react'

<img
  width="960"
  alt=""
  src="https://user-images.githubusercontent.com/378023/197716033-86fdac40-2b14-438c-a712-b003b76dabd3.png"
/>

## Anatomy

> And action bar is mostly a wrapper for different kind of actions. It makes sure alliment and responsiveness works without much configurations.

Reviewer: ```suggestion
And action bar is a wrapper for different kinds of actions. It makes sure alignment and responsiveness works without much configurations.
```

<img
  width="960"

>   alt="A diagram of an action bar with a few buttons, a divider and at the end a button that opens an overflow menu."

Reviewer: ```suggestion
  alt="A diagram of an action bar with a few buttons, a divider and a button that opens an overflow menu at the end."
```

Author: Wouldn't this suggest that the "overflow menu" is opened at the end? Maybe I'll just remove `at the end` altogether to not cause confusion.

  src="https://user-images.githubusercontent.com/378023/198211386-3e48eeb4-8cc6-41c0-85dd-1596aaf0d3ab.png"
/>

### Items

<Box
  as="figure"
  display={["block", "block", "block", "block", "grid"]}
  gridTemplateColumns="1fr 1fr"
  gridGap={7}
  m={0}
  mb={3}
>
  <Text as="p" mt="0">

> Each Action bar item is componsed of a visual and a label.

Reviewer: ```suggestion
Each action bar item is composed of a visual and/or a label.
```

It could have no visual and just text or no label, right?

Author: Hmm.. this is tricky. Each item always needs a label **AND** a visual. But it's true that sometimes the label is only used for the tooltip or the visual is only used when inside the overflow menu. I'll change it to `requires`. Then it's not tied to being visible at all times.

> - `visual`: An icon to best represent the item

Reviewer: ```suggestion
- `visual`: An icon to represent the action
```

- `label`: Text describing the action of the item

> Note that depending on the variant used, the visual or label might not be visible.

Reviewer: ```suggestion
```

If you change it `and/or` above you would not need this.

  </Text>
  <Box>
    <img
      src="https://user-images.githubusercontent.com/378023/198209541-7e647879-2096-46a3-b7bd-6d390ff72e41.png"
      role="presentation"
      width="456"
    />
  </Box>
</Box>

### Dividers

<Box
  as="figure"
  display={["block", "block", "block", "block", "grid"]}
  gridTemplateColumns="1fr 1fr"
  gridGap={7}
  m={0}
  mb={3}
>
  <Text as="p" mt="0">

>     Use a divider to visually seperate items of the same type. No divider is needed between icons and buttons since they are already distinctive enough.

Reviewer: ```suggestion
    Use a divider to visually separate groups of items of the same type. No divider is needed between icons and buttons since they are already distinctive enough.
```

Author: Hmmm.. I think "groups of items of the same type" is not really accurate. A divider should only be used between items of the same type (e.g between icons or between buttons) but not between groups of the same type (e.g. between an icon group and a button group)

<img src="https://user-images.githubusercontent.com/378023/198211711-acca35ec-03d3-4ce1-9d02-05d65909e62f.png" role="presentation" width="456" />

  </Text>
  <Box>
    <img
      src="https://user-images.githubusercontent.com/378023/198211711-acca35ec-03d3-4ce1-9d02-05d65909e62f.png"
      role="presentation"
      width="456"
    />
  </Box>
</Box>

### Overflow menu

When action bar items don't fit in the available space, an overflow button ("kebab" icon) is added at the end of the action bar signaling that there are more actions available. Clicking on the overflow button opens the overflow menu with the remaining actions that didn't fit.

<img
  src="https://user-images.githubusercontent.com/378023/198549383-0e5f42fc-a456-481c-9c44-b9fd6202b70f.png"
  alt=""
  width="960"
/>

#### Sorting

> Actions that don't fit are added to the top of the menu. Meaning that the last action in the action bar will also be the last action when inside the menu. Make sure to add the most important actions to the front since they will be hidden last.

Reviewer: ```suggestion
Actions that don't fit are added to the top of the menu. Meaning that the last action in the action bar will also be the last action inside the menu. Make sure to add the most important actions to the front since they will be hidden last.
```

/>
<Caption>Overflow button appears when not enough space and resizing the action bar updates the overflow menu.</Caption>

## Options

### Gap

> Action bar items use a `8px` gap by default. There is also a more spacious `12px` option available. Note that there is no gap between icons as they already have enough padding and no border.

Reviewer: ```suggestion
Action bar items use an `8px` gap by default. There is also a more spacious `12px` option available. Note that there is no gap between icons as they already have padding and no border.
```

<img
  src="https://user-images.githubusercontent.com/378023/198517276-cf43dc64-dd8f-4af1-b6b0-133c211d99a4.png"
  role="presentation"
  width="960"
/>

### Size

An action bar can have 3 different sizes: Small (`28px`), default (`32px`) and large (`40px`).

<img
  src="https://user-images.githubusercontent.com/378023/198517281-6e615d34-ca51-4a56-8d49-888d09bce426.png"
  role="presentation"
  width="960"
/>

### Max visible

> Sometimes there would be enough space to show all action bar items. But to not overwhelm users with too many actions, an action bar can be configured to limit the amount of items that are visible. In the example below the `maxVisible` is set to `2`. So the rest of the actions are available from the overflow menu.

Reviewer: ```suggestion
An action bar can be configured to limit the amount of items that are visible, even if more enough space is available. This helps to not overwhelm users with too many actions. In the example below the `maxVisible` is set to `2`. So the rest of the actions are available from the overflow menu.
```

<img
  src="https://user-images.githubusercontent.com/378023/198553327-e0e596ab-ddd7-4485-9203-480958dd581c.png"
  role="presentation"
  width="960"
/>

### Item variant

> An Action bar item can be configured with different variants to adjust its rendering based on the environment or use case.

Reviewer: ```suggestion
An action bar item can be configured with different variants to adjust its rendering based on the environment or use case.
```

<DoDontContainer>
  <Do>
    <img
      src="https://user-images.githubusercontent.com/378023/193506398-6d6da18a-b70d-4cd4-b6e4-fd472a24934f.png"
      role="presentation"
      width="456"
    />
    <Caption>Only use "invisible" icon buttons.</Caption>
  </Do>
  <Dont>
    <img
      src="https://user-images.githubusercontent.com/378023/197961141-fe53a86e-e0d9-47ae-ab38-ee95b3f7da1e.png"
      role="presentation"
      width="456"
    />
    <Caption>Don't use other icon button variants.</Caption>
  </Dont>
</DoDontContainer>

#### Button

When there are only 2-3 actions and enough space, consider using the `button` variant. For destructive actions the `dangerButton` variant should be used.

<DoDontContainer>
  <Do>
    <img
      src="https://user-images.githubusercontent.com/378023/198218571-fa10363c-157c-482a-8741-33813f60590e.png"
      role="presentation"
      width="456"
    />
    <Caption>Use default or danger buttons when space allows.</Caption>
  </Do>
  <Dont>
    <img
      src="https://user-images.githubusercontent.com/378023/198218602-549746b6-77c9-4f7b-a758-6a3a962847ae.png"
      role="presentation"
      width="456"
    />

>     <Caption>Don't use other variants. Primary buttons can still be used next to an action bar, just not part of it.</Caption>

Reviewer: ```suggestion
    <Caption>Don't use other variants. Primary buttons can still be used next to an action bar, just not within it.</Caption>
```

  </Dont>
</DoDontContainer>

#### Button with icon

Use the `buttonWithIcon` or `dangerButtonWithIcon` variants to add leading visuals to buttons.

<DoDontContainer>
  <Do>
    <img
      src="https://user-images.githubusercontent.com/378023/198222542-e7d63612-106a-4925-aa45-46ece6df9b7f.png"
      role="presentation"
      width="456"
    />
    <Caption>Use leading visuals to emphasis an action's meaning.</Caption>
  </Do>
  <Dont>
    <img
      src="https://user-images.githubusercontent.com/378023/198222589-3eff32b1-841a-4ebe-b68d-722b39aa2dbc.png"
      role="presentation"
      width="456"
    />
    <Caption>If one of the buttons has a leading visual, then all other buttons should have one too.</Caption>
  </Dont>
</DoDontContainer>

#### Icons and buttons mixed

> It's also ok to mix icons and buttons.

Reviewer: ```suggestion
It's ok to mix icons and buttons.
```

<DoDontContainer>
  <Do>
    <img
      src="https://user-images.githubusercontent.com/378023/198227074-563182c4-063d-4c97-bbc5-8c8fc2ed334f.png"
      role="presentation"
      width="456"
    />

>     <Caption>Use buttons to make them stand out more against the rest of the icons.</Caption>

Reviewer: ```suggestion
    <Caption>Use buttons to make an action stand out more against the rest of the icons.</Caption>
```

  </Do>
  <Dont>
    <img
      src="https://user-images.githubusercontent.com/378023/198227092-797e8b83-491b-483d-ad08-5ffd9c641e2a.png"
      role="presentation"
      width="456"
    />
    <Caption>Don't add leading visuals to buttons if the action bar also has icons.</Caption>
  </Dont>
</DoDontContainer>

## Layout

> Action bars can be used inline next to other content or also full width taking up the entire space.

Reviewer: ```suggestion
Action bars can be used inline next to other content or take up the entire available space.
```

<!-- TODO: Add video -->

### Tooltips

> When hovering over a button, a tooltip will appear that describes the action.

Reviewer: ```suggestion
When hovering over an action, a tooltip with a description appears.
```
